### PR TITLE
bitsetsort peformance check

### LINKED ISSUE #34115: Consider checking out bitsetsort performance
[BitSetSort](https://github.com/minjaehwang/bitsetsort) claims to perform better than pdqsort and std::sort for randomized inputs. I thought to open this issue to see if we can benefit from it.

### PR DESCRIPTION
**Pull request by kitaisreal**
Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Follow up for https://github.com/ClickHouse/ClickHouse/pull/34145.
Closes #34115.

### CODE REVIEW

File: base/base/BitSetSort.h [hunk 1]
```diff
@@ -0,0 +1,854 @@
...
+  * Bitset Sort is a variant of quick sort, specifically BlockQuickSort.
+  * Bitset Sort uses a carefully written partition function to let the compiler generates
+  * SIMD instructions without actually writing SIMD intrinsics in the loop.
+  * Bitset Sort is 3.4x faster (or spends 71% less time) than libc++ std::sort when sorting uint64s and 1.58x faster (or spends 37% less time)
+  * when sorting std::string.
+  * Bitset Sort uses multiple techniques to improve runtime performance of sort. This includes sorting networks,
+  * a variant of merge sort called Bitonic Order Merge Sort that is faster for small N, and pattern recognitions.
+  */
+
+#pragma clang diagnostic ignored "-Wreserved-identifier"
```
Comment (alexey-milovidov) [inline base/base/BitSetSort.h:23]: We need to unset this pragma at the end of file with push/pop.

Comment (alexey-milovidov) [reply base/base/BitSetSort.h:23]: Or include this header with SYSTEM in CMakeLists.

File: base/base/sort.h [hunk 2]
```diff
@@ -30,13 +30,13 @@ void partial_sort(RandomIt first, RandomIt middle, RandomIt last, Compare compar
...
 {
-    ::pdqsort(first, last, compare);
+    ::stdext::bitsetsort(first, last, compare);
 }
 
 template <typename RandomIt>
 void sort(RandomIt first, RandomIt last)
 {
     using value_type = typename std::iterator_traits<RandomIt>::value_type;
     using comparator = std::less<value_type>;
```
Comment (danlark1) [inline base/base/sort.h:40]: I highly recommend to randomize the range in sanitizers and debug mode, same for partial_sort and nth_element

Maybe in a separate PR

**Comment by kitaisreal** [comment]
@Mergifyio update

**Comment by kitaisreal** [comment]
ClickHouse crashes during `00915_simple_aggregate_function` and `00605_intersections_aggregate_functions` produce wrong result, this is related to issue in bitsetsort, need to double check if it is an issue https://github.com/minjaehwang/bitsetsort/pull/3.
After applied fix, `00915_simple_aggregate_function ` do not crash ClickHouse and `00605_intersections_aggregate_functions ` is fine. Lets see other tests results.

CC: @alexey-milovidov

**Review by alexey-milovidov — APPROVED** [review]

**Comment by kitaisreal** [comment]
There are only performance improvements on random data. No performance improvements on read data. On some tests for real dataset test.hits, bitsetsort is slower than pdqsort.
https://s3.amazonaws.com/clickhouse-test-reports/34175/45fa68b4a890b1bda979cadfbd8968422fcecd9d/performance_comparison__actions__[1/4]/report.html
https://s3.amazonaws.com/clickhouse-test-reports/34175/45fa68b4a890b1bda979cadfbd8968422fcecd9d/performance_comparison__actions__[2/4]/report.html
https://s3.amazonaws.com/clickhouse-test-reports/34175/45fa68b4a890b1bda979cadfbd8968422fcecd9d/performance_comparison__actions__[3/4]/report.html

**Pull request merged**